src/cache.rs

```diff
@@ -34,29 +83,110 @@ impl BlockTxIDsCache {
         F: FnOnce() -> Result<Vec<Sha256dHash>>,
     {
         if let Some(txids) = self.map.lock().unwrap().get(blockhash) {
-            self.hits.inc();
             return Ok(txids.clone());
         }
 
-        self.misses.inc();
         let txids = load_txids_func()?;
-        self.map.lock().unwrap().put(*blockhash, txids.clone());
+        let byte_size = 32 /* hash size */ * (1 /* key */ + txids.len() /* values */);
+        self.map
+            .lock()
+            .unwrap()
+            .put(*blockhash, txids.clone(), byte_size);
         Ok(txids)
     }
 }
 
+pub struct TransactionCache {
+    // Store serialized transaction (should use less RAM).
+    map: Mutex<SizedLruCache<Sha256dHash, Vec<u8>>>,
+}
+
+impl TransactionCache {
+    pub fn new(bytes_capacity: usize, metrics: &Metrics) -> TransactionCache {
+        let lookups = metrics.counter_vec(
+            MetricOpts::new(
+                "electrs_transactions_cache",
+                "# of cache lookups for transactions",
+            ),
+            &["type"],
+        );
+        TransactionCache {
+            map: Mutex::new(SizedLruCache::new(bytes_capacity, lookups)),
+        }
+    }
+
+    pub fn get_or_else<F>(&self, txid: &Sha256dHash, load_txn_func: F) -> Result<Transaction>
+    where
+        F: FnOnce() -> Result<Vec<u8>>,
+    {
+        match self.map.lock().unwrap().get(txid) {
+            Some(serialized_txn) => Ok(deserialize(&serialized_txn).unwrap()),
+            None => {
+                let serialized_txn = load_txn_func()?;
+                let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
+                let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
+                self.map
+                    .lock()
```

Comment:
After testing, I found that there is recursive locking happening here and the software hangs.

This patch fixed it for me:
```
--- a/src/cache.rs
+++ b/src/cache.rs
@@ -120,18 +120,17 @@ impl TransactionCache {
         F: FnOnce() -> Result<Vec<u8>>,
     {
         match self.map.lock().unwrap().get(txid) {
-            Some(serialized_txn) => Ok(deserialize(&serialized_txn).unwrap()),
-            None => {
-                let serialized_txn = load_txn_func()?;
-                let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
-                let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
-                self.map
-                    .lock()
-                    .unwrap()
-                    .put(*txid, serialized_txn, byte_size);
-                Ok(txn)
-            }
+            Some(serialized_txn) => return Ok(deserialize(&serialized_txn).unwrap()),
+            None => {}
         }
+        let serialized_txn = load_txn_func()?;
+        let txn = deserialize(&serialized_txn).chain_err(|| "failed to parse cached tx")?;
+        let byte_size = 32 /* key (hash size) */ + serialized_txn.len();
+        self.map
+            .lock()
+            .unwrap()
+            .put(*txid, serialized_txn, byte_size);
+        Ok(txn)
     }
```

Reply:
Many thanks for finding and fixing the issue!
Force-pushed https://github.com/romanz/electrs/pull/172/commits/51d0bf4fb77eb7d8354e3ce2f6de7a9b0c6dcb54.